UL4.1, UL4.2, UL4.3, UL4.4

krystian-hebel opened the pull request.

krystian-hebel committed:
CustomShellPkg: changed GUID for automatic boot to shell

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
PlatformPkgGcc.fdf, PlatformPkgGccIA32.dsc: use CustomShellPkg

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
CustomShellPkg: start memory-mapped application before main shell loop

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
CustomShellPkg: update paths to point to CustomShellPkg

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
CustomShellPkg: use Pcd instead of hardcoded values

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

miczyg1 commented:
@krystian-hebel i guess wrong target branch. 17k lines of code

krystian-hebel commented:
@miczyg1 that is correct. Most of this is copy of ShellPkg.

miczyg1 commented:
@krystian-hebel we should distinguish features/branches. Shell and trampoline stuff maybe should land in boot_to_shell? arbitrary_binary is intended to store code showing integration of arbitrary binaries in SPI flash. If You think that arbitrary_binary is lacking some parts of this change please do the rebase

krystian-hebel commented:
@miczyg1 trampoline requires that binary is present in image.

krystian-hebel committed:
PlatformPkgGcc.fdf: include binary in firmware image for trampoline

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel commented:
@miczyg1 should be fixed now. I wasn't expecting reserved sections in custom_shell.

miczyg1 commented on Vlv2TbltDevicePkg/PlatformPkgGcc.fdf:
```diff
@@ -160,10 +160,11 @@ FV = FVMAIN_COMPACT
-  0x00
-}  # "RESERVED" in HEX
+FILE = hello_world.efi
+#DATA = {
```
Since this branch is intended to launch target binaries via trampoline, these DATA sections are not necessary

miczyg1 commented on Vlv2TbltDevicePkg/PlatformPkgGcc.fdf:
```diff
@@ -160,10 +160,11 @@ FV = FVMAIN_COMPACT
-  0x52, 0x45, 0x53, 0x45, 0x52, 0x56, 0x45, 0x44,
-  0x00
-}  # "RESERVED" in HEX
+FILE = hello_world.efi
```
Let's name it `target_binary.efi` maybe

pietrushnic commented:
@krystian-hebel modified files requires correct copyrights.

pietrushnic commented on Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:
```diff
@@ -314,6 +315,219 @@ InternalEfiShellStartCtrlSMonitor(
 }
 
+/**
+  internal worker function to load and run an image via memory-mapped device path.
```
@krystian-hebel please be consistent: either you use fits bit letter either not.

pietrushnic commented on Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:
```diff
@@ -314,6 +315,219 @@ InternalEfiShellStartCtrlSMonitor(
+                            returned by the command is pointed to by StatusCode.
+  @retval EFI_INVALID_PARAMETER The parameters are invalid.
+  @retval EFI_OUT_OF_RESOURCES Out of resources.
+  @retval EFI_UNSUPPORTED   Nested shell invocations are not allowed.
```
@krystian-hebel maybe correct alignment

pietrushnic commented on Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:
```diff
@@ -314,6 +315,219 @@ InternalEfiShellStartCtrlSMonitor(
+  }
+
+  InitializeListHead(&OrigEnvs);
+  ZeroMem(&ShellParamsProtocol, sizeof(EFI_SHELL_PARAMETERS_PROTOCOL));
```
@krystian-hebel there is no returned values from above functions?

krystian-hebel replied:
No, these are `VOID`s

pietrushnic commented on Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:
```diff
@@ -314,6 +315,219 @@ InternalEfiShellStartCtrlSMonitor(
+  //
+  Status = gBS->LoadImage(
+    FALSE,
+    *ParentImageHandle,
```
@krystian-hebel there is no sanity checking for external parameters?

krystian-hebel replied:
LoadImage is foolproof and does enough checks.

pietrushnic reviewed (changes requested):
Didn't finished review yet, but wanted to provide at least this feedback.

krystian-hebel committed:
PlatformPkgGcc.fdf: change name of binary file, add copyrights

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
CustomShellPkg: fix comments formatting

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel commented:
I've rebased this branch (again) to get rid of unnecessary commits with reserved sections of SPI.

pietrushnic commented on Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
+  CHAR16                        *NewCmdLine;
+
+  if (ParentImageHandle == NULL) {
+    return (EFI_INVALID_PARAMETER);
```
Why we are not consistent?

```suggestion
    return EFI_INVALID_PARAMETER;
```

pietrushnic commented on Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
+    }
+
+    ASSERT(LoadedImage->LoadOptionsSize == 0);
+    if (NewCmdLine != NULL) {
```
@krystian-hebel this check seems to be redundant? You already check if NewCmdLine is NULL right after getting this memory. I don't see why you do it it again.

pietrushnic commented on Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
+  }
+
+  for (Walker = NewCmdLine; Walker != NULL && *Walker != CHAR_NULL ; Walker++) {
+    if (*Walker == L'^' && *(Walker+1) == L'#') {
```
@krystian-hebel why we look for `^#` ? Can you add comment here?

pietrushnic commented on Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
+      ShellParamsProtocol.Argv[0] = ImagePath;
+    }
+
+    Status = gBS->InstallProtocolInterface(&NewHandle, &gEfiShellParametersProtocolGuid, EFI_NATIVE_INTERFACE, &ShellParamsProtocol);
```
@krystian-hebel this invocation has inconsistent style in comparison to other calls.

pietrushnic commented on Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
+    Status = gBS->InstallProtocolInterface(&NewHandle, &gEfiShellParametersProtocolGuid, EFI_NATIVE_INTERFACE, &ShellParamsProtocol);
+    ASSERT_EFI_ERROR(Status);
+
+    ///@todo initialize and install ShellInterface protocol on the new image for compatibility if - PcdGetBool(PcdShellSupportOldProtocols)
```
@krystian-hebel when we plan to implement that? If trampoline is fully functional without it I would recommend to put that into future improvements section.

pietrushnic commented on Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
+    // now start the image and if the caller wanted the return code pass it to them...
+    //
+    if (!EFI_ERROR(Status)) {
+      StartStatus      = gBS->StartImage(
```
```suggestion
      StartStatus = gBS->StartImage(
```

pietrushnic replied:
@krystian-hebel ?

krystian-hebel replied:
Ups, missed it.

pietrushnic commented on Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
+                            );
+      ASSERT_EFI_ERROR(CleanupStatus);
+
+      goto FreeAlloc;
```
@krystian-hebel I have concerns that those `gotos` make code convoluted and quite hard to follow.

pietrushnic replied:
@krystian-hebel do we plan any refactor?

krystian-hebel replied:
I'm not planning to change this. In my opinion repeating cleanup operations after every point of possible error in this code would make unnecessarily deep `if` branches.

pietrushnic commented on Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
+          FreePool (ShellParamsProtocol.Argv[Index]);
+        }
+      }
+      FreePool (ShellParamsProtocol.Argv);
```
@krystian-hebel convention for function calls is not consistent sometimes you have `foo (bar)` and sometimes `foo(bar)`. Coding style use first method.

pietrushnic commented on Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:
```diff
@@ -339,6 +555,22 @@ UefiMain (
+  MemPath[0].Header.Type = HARDWARE_DEVICE_PATH;
+  MemPath[0].Header.SubType = HW_MEMMAP_DP;
+  MemPath[0].Header.Length[0] = (UINT8)sizeof(MEMMAP_DEVICE_PATH);
+  MemPath[0].Header.Length[1] = (UINT8)(sizeof(MEMMAP_DEVICE_PATH)>> 8);
```
```suggestion
  MemPath[0].Header.Length[1] = (UINT8)(sizeof (MEMMAP_DEVICE_PATH) >> 8);
```

krystian-hebel committed:
CustomShellPkg: code refactoring for memory-mapped binary, fixed offset

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
CustomShellPkg: calculate correct offset in standalone and full build

PCDs are assigned different values when building whole image because of
automatic assignments in FDF file. This is somewhat descripted in section
8.4 in EDK2 Build Specification.
https://www.gitbook.com/download/pdf/book/edk2-docs/edk-ii-build-specification

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
CustomShellPkg: fixed indentation

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>
